### PR TITLE
Force close websocket

### PR DESCRIPTION
**Pull request by jwendell**

### CODE REVIEW

File: okhttp/src/main/java/okhttp3/internal/connection/StreamAllocation.java [hunk 1]
```diff
@@ -201,7 +201,7 @@ private RealConnection findConnection(int connectTimeout, int readTimeout, int w
 
   public void streamFinished(boolean noNewStreams, HttpCodec codec) {
     synchronized (connectionPool) {
-      if (codec == null || codec != this.codec) {
+      if (codec != this.codec) {
```
Comment (JakeWharton) [inline okhttp/src/main/java/okhttp3/internal/connection/StreamAllocation.java]: This seems completely unrelated to the change. Please submit it as a separate PR.

File: okhttp-ws/src/main/java/okhttp3/internal/ws/RealWebSocket.java [hunk 2]
```diff
@@ -166,6 +169,9 @@ public void sendPong(Buffer payload) throws IOException {
 
     try {
       writer.writeClose(code, reason);
+
+      // Wait graciously 5 seconds for server close response before shutdown
+      this.source.timeout().timeout(5, TimeUnit.SECONDS);
```
Comment (JakeWharton) [inline okhttp-ws/src/main/java/okhttp3/internal/ws/RealWebSocket.java]: I'm tempted to re-use the connect timeout value here (think: disconnect timeout?). @swankjesse thoughts?

**Comment by JakeWharton** [comment]
Can you include a test for this behavior?

File: okhttp-ws/src/main/java/okhttp3/internal/ws/RealWebSocket.java (hunk 2 above)
Comment (swankjesse) [inline okhttp-ws/src/main/java/okhttp3/internal/ws/RealWebSocket.java]: What bad thing happens if we close eagerly?

Comment (JakeWharton) [inline okhttp-ws/src/main/java/okhttp3/internal/ws/RealWebSocket.java]: Probably nothing terrible, although I wouldn't be surprised if someone eventually complained.

**Comment by jwendell** [comment]
It's been harder than I thought... Writing a test for this has been a pain. Could you guys shed some light here?

I changed MockWebServer to not reply any message from the client. But the client seems to not disconnect. Only when the server side closes the connection (MockWebServer.shutdown()) the client does its close stuff... I tried to put a delay (Thread.sleep) in the client (the test method itself) so that after 5 seconds the client sends the close() but it didn't work...

**Comment by jimmidyson** [comment]
Can we get this resurrected? I don't mind submitting a new PR with these changes in now that master has moved on a bit?

**Comment by jwendell** [comment]
@jimmidyson Feel free to continue the work I began. Like I said, writing the test was the difficult part for me. The patch itself is working fine.

**Comment by jimmidyson** [comment]
Things have moved on in 3.4 branch now so the patch may be a bit more involved. For me (as I'm sure for you) the issue was due to a bug in Kubernetes 1.3, which is fixed in 1.4 & OpenShift so might just note the incompatibility instead...

**Comment by swankjesse** [comment]
With recent changes to our API I think this is no longer the right place for this behavior to live. In particular we should force the close after writing a close, not after reading it. I’m going to fix the bug, but the codebase has shifted enough since this PR was sent that a new fix makes more sense.

Thanks for sending the PR, and sorry for not getting it in!
